packages/api/cms-api/src/user-permissions/user-permissions.module.ts

```diff
@@ -24,31 +33,56 @@ import { UserPermissionsAsyncOptions, UserPermissionsOptions, UserPermissionsOpt
             provide: CURRENT_USER_LOADER,
             useClass: UserPermissionsCurrentUserLoader,
         },
+        ContentScopeService,
+        {
+            provide: APP_GUARD,
+            useClass: UserPermissionsGuard,
+        },
     ],
-    exports: [CURRENT_USER_LOADER],
+    exports: [CURRENT_USER_LOADER, ContentScopeService],
 })
 export class UserPermissionsModule {
-    static forRoot(options: UserPermissionsOptions): DynamicModule {
+    static forRoot(options: UserPermissionsModuleSyncOptions): DynamicModule {
         return {
             module: UserPermissionsModule,
             providers: [
                 {
                     provide: USER_PERMISSIONS_OPTIONS,
                     useValue: options,
                 },
+                {
+                    provide: USER_PERMISSIONS_USER_SERVICE,
+                    useClass: options.UserService,
+                },
+                {
+                    provide: ACCESS_CONTROL_SERVICE,
+                    useClass: options.AccessControlService ?? AccessControlService,
+                },
             ],
         };
     }
 
-    static forRootAsync(asyncOptions: UserPermissionsAsyncOptions): DynamicModule {
+    static forRootAsync(options: UserPermissionsModuleAsyncOptions): DynamicModule {
         return {
             module: UserPermissionsModule,
-            imports: asyncOptions.imports,
-            providers: [this.createProvider(asyncOptions)],
+            imports: options.imports,
+            providers: [
+                this.createProvider(options),
+                {
+                    provide: USER_PERMISSIONS_USER_SERVICE,
+                    useFactory: (options: UserPermissionsAsyncOptions) => options.userService,
+                    inject: [USER_PERMISSIONS_OPTIONS],
+                },
+                {
+                    provide: ACCESS_CONTROL_SERVICE,
+                    useFactory: (options: UserPermissionsAsyncOptions) => options.accessControlService ?? new AccessControlService(),
```

Comment:
Why do you need this `new` call here, and not in `forRoot`? Isn't this anti Nest DI? 🤔

Reply:
In `forRoot` we use `useClass` because we don't have to care about DI. In `forRootAsync` we have to use `useFactory` because we provide the possibility to inject an instance of `AccessControlService` by the application.

We also provide a fallback for this service and yes, we could solve this with an own provider. But since the default `AccessControlService` does not use DI we just can instantiate it. I see noe breaking of DI at all.